### PR TITLE
Add ShootExtensionStatus controllers

### PR DESCRIPTION
**Pull request by danielfoehrKn**
**How to categorize this PR?**

/area quality
/kind enhancement

**What this PR does / why we need it**:

Contains 3 commits
- Add ShootExtensionStatus to SeedRestriction Admission Webhook
- Add ShootExtensionStatus to SeedAuthorizer Authorization Webhook
- Add federated seed controllers for the ShootExtensionStatus resource.

Part of #3873 

 - Creation of the `ShootExtensionStatus` is done via ShootReconciliation flow. Deletion is implicit via owner reference on the particular Shoot resource.
- Adds the reconciliation of the `ShootExtensionStatus` to the federated seed controller in the Gardenlet

Only enables the sync for the `Infrastructure` and `Worker` extension resources for now.
 - This is to avoid unnecessary network traffic. The mentioned resources are the only ones that currently write a `providerStatus`  in their status. 
 - Also, the Gardenlet is unable to detect and therefore `not` sync an "empty" `providerStatus` that only contains `apiVersion` and `kind` such as:
 ```
  providerStatus:
      apiVersion: calico.networking.extensions.gardener.cloud/v1alpha1
      kind: NetworkStatus
```

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator
THe Gardenlet exposes the provider status of extension resources from the Seed cluster in the Garden cluster via the "ShootExtensionStatus" resource.
```

### CODE REVIEW

File: pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+func (s *ShootExtensionStatusControl) patchShootExtensionStatus(ctx context.Context, shootExtensionStatus *gardencorev1alpha1.ShootExtensionStatus, shootExtensionStatusCopy *gardencorev1alpha1.ShootExtensionStatus, kind string) (reconcile.Result, error) {
+	if err := s.gardenClient.Patch(ctx, shootExtensionStatus, client.MergeFromWithOptions(shootExtensionStatusCopy, client.MergeFromWithOptimisticLock{})); err != nil {
+		message := fmt.Sprintf("The %q extension status for Shoot %q was NOT successfully synced: %v", shootExtensionStatus.Name, kind, err)
+		s.log.Error(message)
+		return reconcile.Result{}, err
+	}
+
+	message := fmt.Sprintf("The provider status of the %q extension of Shoot %q was successfully synced to the Garden cluster", kind, shootExtensionStatus.Name)
+	s.log.Info(message)
```
Comment (timebertt) [inline pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go]:
```suggestion
	s.log.Debug(message)
```

File: pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+	}
+
+	if shoot == nil {
+		return nil, fmt.Errorf("cluster resource %s doesn't contain shoot resource", cluster.Name)
+	}
+	return shoot, nil
+}
+
+func (s *ShootExtensionStatusControl) patchShootExtensionStatus(ctx context.Context, shootExtensionStatus *gardencorev1alpha1.ShootExtensionStatus, shootExtensionStatusCopy *gardencorev1alpha1.ShootExtensionStatus, kind string) (reconcile.Result, error) {
+	if err := s.gardenClient.Patch(ctx, shootExtensionStatus, client.MergeFromWithOptions(shootExtensionStatusCopy, client.MergeFromWithOptimisticLock{})); err != nil {
```
Comment (timebertt) [inline pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go:181]: This has a high chance of conflicting with other current updates to the `ShootExtensionStatus`.
Have you considered a strategic merge patch without optimistic locking?
This will need some `patchMergeKey` here: https://github.com/gardener/gardener/blob/abfb062f5121cf53b7e52ddf65022c3f6b867bb4/pkg/apis/core/v1alpha1/types_shootextensionstatus.go#L34

File: pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+			shoot, err := s.getShootForRequest(ctx, req)
+			if err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to delete the provider status of the %q extension (%s/%s) from the ShootExtensionStatus: %w",
+					kind,
+					req.NamespacedName.Namespace,
+					req.NamespacedName.Name,
+					err)
+			}
+
+			shootExtensionStatus := &gardencorev1alpha1.ShootExtensionStatus{}
```
Comment (timebertt) [inline pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go:78]: Can you check how this will play with the `SeedAuthorizer`? 
I guess we can't/shouldn't simply watch all `ShootExtensionStatus`es.

**Review by timebertt — COMMENTED** [review]
I have added some general concerns, haven't checked everything in detail yet.

File: pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go (hunk 2 above)
Comment (danielfoehrKn) [reply pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go:181]: Yeah, I tried using the `strategicMergePatch` but there is an issue with using `rawExtension` field. That is also the reason why in the ShootState a json-merge-patch is used @plkokanov .

It seems like when using `strategicMergePatch`  it also tries to merge the `rawExtension` field - and it does not know how to do that for lists inside the field (instead of simply replacing the field content).

This can be observed consequently only when there is a conflict with an existing list entry.

>  unable to find api field in struct RawExtension for the json field “ec2”

If you know how to use the `strategicMergePatch` safely with `rawExtension` fields, I would of course prefer that.

File: pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go (hunk 3 above)
Comment (danielfoehrKn) [reply pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go:78]: Good point regarding the SeedAuthorizer, though I would propose to take care of the seed restriction admission plugin & authorizer webhook in a follow-up PR.

However, we do not watch `ShootExtensionStatuses` here as far as I can tell (only get request on a `ShootExtensionStatus` in a certain namespace). Or is the controller-runtime opening informer-based watches on all `ShootExtensionStatuses` under the hood?. 
If yes, what can be done about it?

Comment (rfranzke) [reply pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go:78]: > though I would propose to take care of the seed restriction admission plugin & authorizer webhook in a follow-up PR.

At this point, we think only `Secret` handling is missing in the seed authorizer. I plan to open a PR for it next week. After it's open/merge merge, I would like to go ahead and play with Gardener while activating the seed authorizer.
Hence, if we now NOT merge this PR and don't have the necessary code for `ShootExtensionStatus`es in the seed authorizer, I'm blocked. So, please add it to this PR instead of postponing it (it shouldn't be too much of necessary code).

> If yes, what can be done about it?

You might want to look into https://github.com/gardener/gardener/pull/3803.

File: pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go (hunk 1 above)
Comment (danielfoehrKn) [reply pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go]: The ShootState controller also logs with [Info level on success](https://github.com/gardener/gardener/blob/3d5f3f3fd3d65b57f775cb14fb251463ace8b731/pkg/gardenlet/controller/federatedseed/extensions/shootstate_control.go#L135).  Is there a reason you would like to change that?

**Comment by danielfoehrKn** [comment]
Added
 - client does not list & watch anymore for ShootExtensionStatus resources
 - Added handling in Admission & Authorization webhooks in separate PRs

PTAL @rfranzke @timebertt

File: pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go (hunk 1 above)
Comment (timebertt) [reply pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go]: Well, it's just annoying to log such messages on `Info` and unnecessarily spams the logs 😄

File: pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go (hunk 2 above)
Comment (timebertt) [reply pkg/gardenlet/controller/federatedseed/extensions/shootextensionstatus_control.go:181]: Yeah, checked this in detail now and it seems to be very tricky indeed.

I see several problems:
- we would need a `UID` field in `ExtensionStatus` which can be used as a `patchMergeKey` for `ShootExtensionStatus.Statuses` (strategic merge can't handle multiple merge keys, I think)
- we would need to use the [`replace` directive](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/strategic-merge-patch.md#replace-directive) to tell strategic merge patch to not merge but replace the `RawExtension` field. The problem here is, that adding it to the `Status` field (`patchStrategy:"replace"`) is not sufficient, as it will send all statuses on every patch. The directive has to be added inside `RawExtension`'s json. However, then the patch computation will still fail, as the computation doesn't take the directives into account but only the patch strategy.

TL;DR: this is just to confusing to figure out, so let's use the merge patch with optimistic locking as you proposed 👍

**Comment by danielfoehrKn** [comment]
Do you still want to take a look ? @timebertt & @rfranzke

**Comment by timebertt** [comment]
Sorry for the delay, it's quite full here. Will try to recheck today.

File: pkg/admissioncontroller/webhooks/auth/seed/graph/eventhandler_shootextensionstatus.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+import (
+	"time"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	toolscache "k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+)
+
+func (g *graph) setupShootExtensionStatusWatch(informer cache.Informer) {
```
Comment (rfranzke) [inline pkg/admissioncontroller/webhooks/auth/seed/graph/eventhandler_shootextensionstatus.go:25]: I suggest to use a metadata-only watch to reduce the memory footprint of GAC (you seem to be only interested in namespace/name of the object).

File: pkg/gardenlet/controller/federatedseed/extensions/artifact.go [hunk 5]
```diff
@@ -209,6 +230,25 @@ func newStateArtifact(gvk schema.GroupVersionKind, newObjFunc func() client.Obje
 	return a
 }
 
+func newExtensionStatusArtifact(gvk schema.GroupVersionKind, newObjFunc func() client.Object, fn predicateFn) *artifact {
```
Comment (rfranzke) [inline pkg/gardenlet/controller/federatedseed/extensions/artifact.go:233]: It's probably not needed to let gardenlet watch all these objects and return the status "asap" similar to the `ShootState`. The `.status.providerStatus` should only change after reconciliation, hence, it would be sufficient if you add another task to the end of the reconciliation flow which syncs the provider statuses, for example.

File: pkg/gardenlet/controller/federatedseed/extensions/common.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+	"context"
+
+	apiextensions "github.com/gardener/gardener/pkg/api/extensions"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// GetExtensionObject is a helper to get the extension object for a key and objectCreator function
+func GetExtensionObject(ctx context.Context, seedClient client.Client, key types.NamespacedName, objectCreator func() client.Object) (extensionsv1alpha1.Object, error) {
```
Comment (rfranzke) [inline pkg/gardenlet/controller/federatedseed/extensions/common.go:27]: unit test?

File: pkg/extensions/cluster.go [hunk 7]
```diff
@@ -212,3 +214,21 @@ func GetShoot(ctx context.Context, c client.Client, namespace string) (*gardenco
 func NewGardenDecoder() runtime.Decoder {
 	return serializer.NewCodecFactory(gardenScheme).UniversalDecoder()
 }
+
+// ClusterFromRequest returns the cluster resource for a given reconcile.Request
+func ClusterFromRequest(ctx context.Context, seedClient client.Client, req reconcile.Request) (*extensionsv1alpha1.Cluster, error) {
```
Comment (rfranzke) [inline pkg/extensions/cluster.go:219]: unit test?

File: pkg/gardenlet/controller/shoot/shoot_control_reconcile.go [hunk 8]
```diff
@@ -165,7 +169,7 @@ func (c *Controller) runReconcileShootFlow(ctx context.Context, o *operation.Ope
 		deployInfrastructure = g.Add(flow.Task{
 			Name:         "Deploying Shoot infrastructure",
 			Fn:           flow.TaskFn(botanist.DeployInfrastructure).RetryUntilTimeout(defaultInterval, defaultTimeout),
-			Dependencies: flow.NewTaskIDs(deploySecrets, deployCloudProviderSecret, deployReferencedResources),
+			Dependencies: flow.NewTaskIDs(deploySecrets, deployCloudProviderSecret, deployReferencedResources, ensureShootExtensionStatus),
```
Comment (rfranzke) [inline pkg/gardenlet/controller/shoot/shoot_control_reconcile.go:172]: Why is "only" this task depending on `ensureShootExtensionStatus`? I suggest to remove it here because, consequently, you would have to add it to all other tasks creating extension resources as well.

File: pkg/operation/botanist/component/shootextensionstatus/shoot_extension_status.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+		shoot:        shoot,
+	}
+}
+
+type shootExtensionStatus struct {
+	gardenClient client.Client
+	shoot        *gardencorev1beta1.Shoot
+}
+
+func (m *shootExtensionStatus) Deploy(ctx context.Context) error {
```
Comment (rfranzke) [inline pkg/operation/botanist/component/shootextensionstatus/shoot_extension_status.go:50]:
```suggestion
func (s *shootExtensionStatus) Deploy(ctx context.Context) error {
```

File: pkg/operation/botanist/component/shootextensionstatus/shoot_extension_status.go [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+) ShootExtensionStatus {
+	return &shootExtensionStatus{
+		gardenClient: client,
+		shoot:        shoot,
+	}
+}
+
+type shootExtensionStatus struct {
+	gardenClient client.Client
+	shoot        *gardencorev1beta1.Shoot
```
Comment (rfranzke) [inline pkg/operation/botanist/component/shootextensionstatus/shoot_extension_status.go:47]: You only seem to require namespace/name

File: pkg/operation/botanist/component/shootextensionstatus/shoot_extension_status.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+	ownerReference.BlockOwnerDeletion = &blockOwnerDeletion
+
+	_, err := controllerutils.StrategicMergePatchOrCreate(ctx, m.gardenClient, status, func() error {
+		status.OwnerReferences = []metav1.OwnerReference{*ownerReference}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (rfranzke) [inline pkg/operation/botanist/component/shootextensionstatus/shoot_extension_status.go:72]:
```suggestion
	return err
```

File: pkg/operation/botanist/component/shootextensionstatus/shoot_extension_status.go (hunk 9 above)
Comment (rfranzke) [inline pkg/operation/botanist/component/shootextensionstatus/shoot_extension_status.go:50]: Unit test?

Comment (rfranzke) [inline pkg/operation/botanist/component/shootextensionstatus/shoot_extension_status.go:50]: Why is this implemented different than `ShootState`? To me, implementing the component interface seems unnecessary overhead and can probably be avoided.

**Comment by timebertt** [comment]
Not much to add to @rfranzke's comment from my side.

File: pkg/gardenlet/controller/shoot/shoot_control_reconcile.go (hunk 8 above)
Comment (danielfoehrKn) [reply pkg/gardenlet/controller/shoot/shoot_control_reconcile.go:172]: > I suggest to remove it here  

where would you put it instead?

File: pkg/gardenlet/controller/federatedseed/extensions/artifact.go (hunk 5 above)
Comment (danielfoehrKn) [reply pkg/gardenlet/controller/federatedseed/extensions/artifact.go:233]: I cannot follow. This is generic functionality that should be used by other components. How do you know that it is not need to "watch all these objects and return the status "asap"? 
Just to name a few possibilities 
 - admission plugins forbid altering of infrastructure resource during Shoot creation,....
 - Third-party that wants to act upon created infrastructure resources. 
   - Why Let them wait until Shoot is reconciled?
   - What happens if the Infrastructure is already created, but the initial creation of the Shoot fails? 
 - During Shoot deletion: Fails before Shoot can be deleted. Infra resources are already gone but still in the `ShootExtensionStatus`.

Anyways, i mentioned a couple of times in our internal syncs that I will implement it this way.
Plus this PR is now a couple of weeks old and this comment would mean to completely change the implementation + tests.
As this functionality has low priority there are much more important things to be tackled, I consider dropping this PR and revert the previous commit with the unused API.
If this becomes a higher priority in the future we can revisit it.

My lesson for this is, that I ll not accept tasks that are supposed to be done as part of a 5-day quality week that will foreseeably take much longer. This is especially true when adding new APIs that always spark discussions. 
In addition, I would appreciate if these kinds of ideas that have a huge impact on the implementation are communicated if possible in advance.

Comment (rfranzke) [reply pkg/gardenlet/controller/federatedseed/extensions/artifact.go:233]: IIUC, the implementation leverages the already established watches for syncing the `ShootState` in the `federatedseed` controller, correct? In this case (no new watches are created), it seems to be fine to proceed with the implementation as is. You have enlisted a few cases in which immediate-syncing seems beneficial (assuming the existing watches are used).

In retrospect, even if the change seemed to be small enough, a GEP would've been the proper way for proposing this enhancement. #3873 didn't have much activity (rather none) and there was no "main reviewer" assigned to it. Also, I don't think there was a dedicated meeting about this topic and its implementation details. Probably, it cannot be assumed that there is enough bandwidth to provide such "implementation ideas" ad-hoc just because it was mentioned once or twice in a sync while everybody else is busy with other topics (unless there is a "main reviewer", of course).

Anyways, let's proceed now and finish this up. It's not necessary to revert everything. Significant time was already invested and we still would like to have the feature.

**Comment by timuthy** [comment]
@danielfoehrKn as you mentioned the intention to rewrite this functionality, can we close this PR?

**Comment by timuthy** [comment]
I suggest to close this PR if there is no feedback by the end of the week.
